### ISSUE TITLE
0.55 - preview 1 : CollectionView broken

### ISSUE THREAD

**Issue by vshapenko** [fsprojects/Fabulous#733]
Looks like items in CollectionView should be a kind of "special case" in new diff algo, because collection view looks completely broken.  It looks like virtualization breaks out hopes for better world...

**Comment by TimLariviere** [comment]
@vshapenko Can you share a small sample so we can fix this, please?

**Comment by vshapenko** [comment]
Ok, here is little one. Do not know how it is representative, but looks strange: 

```fsharp
 module App = 
    type Model = 
      { Items:int list }

    type Msg = 
        | Next 
    

   

    let initModel () = { Items = [1..100] }

    let init () = initModel () , []

    let update msg model =
        match msg with
        | Next -> model,Cmd.none

    let random = Random(DateTime.UtcNow.Ticks|>int)
    let view (model: Model) dispatch =
        let views =
            [
            
             for item in model.Items do
                 let kind = random.Next(0,100)
                 if(kind<20) then
                    yield View.Label(fontSize=FontSize 40.,text= sprintf "ITEM %i" item,backgroundColor =Color.Aqua)
                    
                 elif (kind<40) then
                     yield View.Frame(
                            cornerRadius = 24.,
                            backgroundColor = Color.Yellow,
                            content=View.Label(fontSize=FontSize 24.,text= sprintf "ITEM %i" item))
                 elif (kind<70) then
                     yield View.Label(fontSize=FontSize 24.,text= sprintf "Item %i" item,backgroundColor=Color.Red)
                 
                 else
                    yield View.Label(fontSize=FontSize (14.),text= sprintf "Item %i" item,backgroundColor=Color.Green) 
                 ]
        View.ContentPage(
          content=View.CollectionView(
            items = views
            
            ))
             
    let program = 
        Program.mkProgram init update view
```

**Comment by vshapenko** [comment]
@TimLariviere , looks like this is a kind of view recyle behaviour, maybe we need to clear all previous attributes on update

**Comment by TimLariviere** [comment]
@vshapenko I'm not seeing any difference between v0.54 and v0.55.0-preview1 using the sample you provided.
Am I missing something?

![versus](https://user-images.githubusercontent.com/6429007/81306761-2178cd00-9080-11ea-8cbd-14241aa90eaa.gif)

**Comment by vshapenko** [comment]
@TimLariviere , hmmm. I will try to dig deeper. Problem is my app has rich ui inside collection view, i will think how to repeat it

**Comment by vshapenko** [comment]
@TimLariviere  here is updated sample with behaviour i do not understand: 

```fsharp
module App = 
    type Model = 
      { Items:int list }

    type Msg = 
        | Next 
    

   

    let initModel () = { Items = [1..20] }

    let init () = initModel () , []

    let update msg model =
        match msg with
        | Next ->
            {
             model with Items = model.Items@[model.Items.Length..model.Items.Length+20]
             }
            ,Cmd.none

    let random = Random(DateTime.UtcNow.Ticks|>int)
    let drawItem item  =
        View.Grid(
                     coldefs =[Dimension.Auto;Dimension.Star;Dimension.Auto],
                     children =
                         [
                             if(item%5=0) then
                                         View.Label(margin=Thickness 16.,horizontalOptions=LayoutOptions.Start,fontSize=FontSize 40.,text= sprintf "ITEM %i aqua" item,backgroundColor =Color.Aqua)
                                
                             elif (item % 3 =0) then
                                  View.Frame(
                                        cornerRadius = 24.,
                                        backgroundColor = Color.Yellow,
                                        content=View.Label( margin=Thickness 16.,horizontalOptions=LayoutOptions.End,fontSize=FontSize 24.,text= sprintf "ITEM %i yellow" item))
                                  |>column 2  
                             else
                                 View.Label(margin=Thickness 16.,fontSize=FontSize (14.),text= sprintf "Item %i green" item,backgroundColor=Color.Green)
                                 |>column 1
                         ]
                 )
    let view (model: Model) dispatch =
        let views = model.Items|>List.map drawItem
           
        View.ContentPage(
          content=View.CollectionView(
            itemSizingStrategy = ItemSizingStrategy.MeasureAllItems,
            items = views,
            remainingItemsThreshold = 5,
            remainingItemsThresholdReached = (fun ()->dispatch Next)
            
            ))
             
    let program = 
        Program.mkProgram init update view

type CounterApp () as app = 
    inherit Application ()

    let runner =
        App.program
        |> Program.withConsoleTrace
        |> XamarinFormsProgram.run app
```

**Comment by vshapenko** [comment]
@TimLariviere , after changing 
ViewElementDataTemplateSelector to
```type ViewElementDataTemplateSelector() =
    inherit DataTemplateSelector()
    let cache = Dictionary<Type, ViewElementDataTemplate>()
    override this.OnSelectTemplate(item, _) =
        let holder = item :?> IViewElementHolder
        let targetType = holder.ViewElement.TargetType
        let template = ViewElementDataTemplate(targetType)
        cache.[targetType] <- template
        template :> DataTemplate```
problem gone.

**Comment by Dolfik1** [comment]
There are some exceptions when using the key attribute in CollectionView elements. Try to add key attribute to each collection item and it cause exception: ArgumentOutOfRangeException in `move` function.

I fixed it with this function:
```fsharp
let inline safeInsert index item (list: IList<'a>) =
  if list.Count > index then
    list.Insert(index, item)
  else
    list.Add(item)
```

But then I got another exception (key already added) in this code:
```fsharp
keyedElements.Add(key, prevChild)
```

**Comment by vshapenko** [comment]
@TimLariviere , could you please take a look at that issue, please?

**Comment by TimLariviere** [comment]
@vshapenko Could you upload a video of the exact behavior you're observing so I can understand the real issue you're encountering, please?

I'm seeing choppy scrolling, and freezes when reaching the end of the list, is that it?

The choppy scrolling might be because of items of different heights.
Virtualization works smoothly only when dealing with same height items (since XF can predict the list size by doing ItemsCount * ItemHeight).

For the freeze, need to investigated deeper.

But I'm thinking the way ListView/CollectionView are used in Fabulous today might not be the best when dealing with virtualization.

From a virtualization standpoint, it would be better to do
```
View.CollectionView(
    items = seq { 0 .. 1000000 },
    itemTemplate = (fun item ->
        if item % 2 = 0 then View.Label()
        elif item % 2 = 1 then View.Button()
        else View.Image()
    )
)
```

**Comment by vshapenko** [comment]
Ok, here is video : 
[device-2020-05-18-194845.zip](https://github.com/fsprojects/Fabulous/files/4645468/device-2020-05-18-194845.zip)

**Comment by vshapenko** [comment]
> @vshapenko Could you upload a video of the exact behavior you're observing so I can understand the real issue you're encountering, please?
> 
> I'm seeing choppy scrolling, and freezes when reaching the end of the list, is that it?
> 
> The choppy scrolling might be because of items of different heights.
> Virtualization works smoothly only when dealing with same height items (since XF can predict the list size by doing ItemsCount * ItemHeight).
> 
> For the freeze, need to investigate deeper.
> 
> But I'm thinking the way ListView/CollectionView are used in Fabulous today might not be the best when dealing with virtualization.
> 
> From a virtualization standpoint, it would be better to do
> 
> ```

the problem is not in scrolling itself. Problem is in drawing same item in different way.

**Comment by vshapenko** [comment]
@TimLariviere , and here is sample app for video above:

```fsharp
module App = 
    type Model = 
      { Item:int
        Items:int list
        }

    type Msg = 
        | Next 
    

   

    let random = Random(DateTime.UtcNow.Ticks|>int)
    
    let initModel () = {Item =random.Next(50);Items = [] }

    let init () = initModel () , []

    let update msg model =
        match msg with
        | Next ->
            let newItems = [
                            yield! model.Items
                            for i in 0..50 do
                                yield random.Next(50)

                           ]
            {model with Items= newItems}
            ,Cmd.none

    
    let drawItem item  =
         let firstVisible = (item%5=0)
         let secondVisible = item % 3=0
         let thirdVisible = not (firstVisible || secondVisible)
         View.Grid(
                     rowSpacing =0.,
                     columnSpacing =0.,
                     coldefs =[Dimension.Auto;Dimension.Star;Dimension.Auto],
                     rowdefs = [Dimension.Auto],
                     children =
                         [
                             
                                  if(firstVisible) then
                                    yield  View.Label(horizontalOptions=LayoutOptions.Start,fontSize=FontSize 20.,text= sprintf "ITEM %i aqua" item,backgroundColor =Color.Aqua,isVisible=firstVisible)

                                  if(secondVisible) then
                                      yield View.Frame(
                                            margin=Thickness 0.,
                                            cornerRadius = 24.,
                                            backgroundColor = Color.Yellow,
                                            content=View.Label(margin=Thickness 16.,horizontalOptions=LayoutOptions.End,fontSize=FontSize 22.,text= sprintf "ITEM %i yellow" item)
                                        )|>column 2  
                             
                                  if(thirdVisible) then 
                                   yield View.Label(fontSize=FontSize (14.),text= sprintf "Item %i green" item,backgroundColor=Color.Green) |>column 1
                         ]
                 )
    let view (model: Model) dispatch =
           
        View.ContentPage(
          content=View.StackLayout(
            children = [
                View.CollectionView(
                                       itemSizingStrategy =ItemSizingStrategy.MeasureAllItems,
                                       items = (model.Items|>List.map drawItem)
                                   ) |> verticalOptions LayoutOptions.StartAndExpand
               
                View.Button(text ="Next", command = (fun()->dispatch Next))
                
              ]
            )
           
            
            )
             
    let program = 
        Program.mkProgram init update view
```

**Comment by TimLariviere** [comment]
@vshapenko Thanks. I can now reproduce.
Weirdly enough, it seems to impact Android only. iOS seems fine.
For some reasons, when CollectionView recycles a row, it has the wrong previous ViewElement and so when the diffing happens, the actual row is in an invalid state.

Need to find why the previous ViewElement is not correct.

**Comment by TimLariviere** [comment]
>Looks like items in CollectionView should be a kind of "special case" in new diff algo, because collection view looks completely broken.

@vshapenko The more I think about it, the more I agree with you on making ListView/CollectionView special.
Fabulous is currently fighting against the virtualization and row recycling of Xamarin.Forms, so ultimately it can't work properly.

I think we should use the `updateCollectionGeneric` function only for the layout controls (Grid, StackLayout, etc) and not the items controls (ListView / CollectionView).

`updateCollectionGeneric` reuse/reordering is too aggressive for the items controls. Xamarin.Forms already does a lot by itself.
We should reorder only when the key changes or we reuse the same ViewElement (same instance), but only reassign the new ViewElement when a row can be reused.
It's a lot like what was done previously.

**Comment by vshapenko** [comment]
@TimLariviere , i think, we may have an old diff for collections, but with addition of key. From my PoV, to make things work in case of virtualization, we should keep keyed controls somewhere for later reuse. I mean scenario when control is out of view(scrolled down), but we want to reuse it. Sample of such control is Map, for example, which has long initialization state, but needs persistence all the way long. I do not know yet how to implement it properly.

**Comment by TimLariviere** [comment]
>Sample of such control is Map, for example, which has long initialization state, but needs persistence all the way long. I do not know yet how to implement it properly.

The problem with that is that you're completely replacing the virtualization process of Xamarin.Forms with your own. So I don't think it will be possible.

**Comment by TimLariviere** [comment]
Should be fixed in `0.55.0-preview3`
